R/Rules-methods.R

```diff
@@ -526,259 +521,136 @@ setMethod(
-          probTarget[as.numeric(names(tab))] <- tab
-          probTarget <- probTarget / nrow(biomLevelSamples)
-        }
-      } else {
-        ## otherwise target is absolute
-
-        ## For each sample, look which dose is maximizing the
-        ## simultaneous probability to be in the target biomarker
-        ## range and below overdose toxicity
-        probTarget <- numeric(ncol(biomLevelSamples))
-        probTarget <- sapply(
-          seq(1, ncol(biomLevelSamples)),
-          function(x) {
-            sum(biomLevelSamples[, x] >= nextBest@target[1] &
-              biomLevelSamples[, x] <= nextBest@target[2]) /
-              nrow(biomLevelSamples)
-          }
-        )
-      }
-
-      ## Now compute probabilities to be in
-      ## overdose tox interval
-      probOverdose <-
-        colMeans((probSamples > nextBest@overdose[1]) &
-          (probSamples <= nextBest@overdose[2]))
-
-      ## which doses are eligible after accounting
-      ## for maximum possible dose and discarding overdoses?
-      dosesBelowLimit <-
-        if (length(doselimit)) {
-          (data@doseGrid <= doselimit)
-        } else {
-          rep(TRUE, length(data@doseGrid))
-        }
-
-      dosesOK <- which(dosesBelowLimit &
-        (probOverdose < nextBest@max_overdose_prob))
-
-      ## check if there are doses that are OK
-      if (length(dosesOK)) {
-        ## For placebo design, if safety allow, exclude placebo from
-        ## the recommended next doses
-        if (data@placebo & (length(dosesOK) > 1L)) {
-          dosesOK <- dosesOK[-1]
-        }
-
-        ## what is the recommended dose level?
-
-        ## if maximum target probability is higher than the numerical
-        ## threshold, then take that level, otherwise stick to the
-        ## maximum level that is OK:
-        doseLevel <-
-          if (max(probTarget[dosesOK]) > nextBest@target_thresh) {
-            which.max(probTarget[dosesOK])
-          } else {
-            which.max(data@doseGrid[dosesOK])
-          }
+#' @describeIn nextBest find the next best dose based on the dual endpoint
+#'   model. The additional list element `probs` contains the target and
+#'   overdosing probabilities (across all doses in the dose grid) used in the
+#'   derivation of the next best dose.
+#'
+#' @aliases nextBest-NextBestDualEndpoint
+#'
+#' @export
+#' @example examples/Rules-method-nextBest-NextBestDualEndpoint.R
+#'
+setMethod(
+  f = "nextBest",
+  signature = signature(
+    nextBest = "NextBestDualEndpoint",
+    doselimit = "numeric",
+    samples = "Samples",
+    model = "DualEndpoint",
+    data = "Data"
+  ),
+  definition = function(nextBest, doselimit, samples, model, data, ...) {
+    doselimit <- ifelse(missing(doselimit) || length(doselimit) == 0, Inf, doselimit)
 
-        ret <- data@doseGrid[dosesOK][doseLevel]
+    # Biomarker samples at the dose grid points.
+    biom_samples <- samples@data$betaW
+
+    if (nextBest@target_relative) {
```

Comment:
I guess the result of this if/else blocks is `prob_target`, therefore would recommend to use the pattern
```r
prob_target <- if () else .. etc.
```

Reply:
Yap, absolutely, thanks I just missed that.